[media-controls] Allow for Slider subclasses

graouts opened the pull request:
#### b6fe39f037b14ed59830c0e143549868f549f9bb
<pre>
[media-controls] Allow for Slider subclasses
<a href="https://bugs.webkit.org/show_bug.cgi?id=241973">https://bugs.webkit.org/show_bug.cgi?id=241973</a>

Reviewed by Dean Jackson.

Split the existing code between a SliderBase superclass and a Slider concrete subclass.
SliderBase takes care of defining the model, hosting the &lt;input&gt; and registering the required
DOM events on it, while the subclass takes care of defining its rendering.

* LayoutTests/media/modern-media-controls/resources/media-controls-loader.js:
* LayoutTests/media/modern-media-controls/slider/slider-constructor-expected.txt:
* LayoutTests/media/modern-media-controls/slider/slider-constructor.html:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/modern-media-controls/controls/slider-base.css: Added.
(.slider):
(.slider &gt; .appearance,):
(.slider &gt; .appearance):
(.slider.allows-relative-scrubbing &gt; .appearance):
(.slider.allows-relative-scrubbing &gt; .appearance.changing):
(.slider &gt; input):
(.slider.allows-relative-scrubbing &gt; input):
(.ios .slider &gt; input):
(.ios .slider:not(.allows-relative-scrubbing) &gt; input):
(.slider &gt; input::-webkit-slider-thumb):
(.slider.disabled &gt; input):
(.ios .slider &gt; input::-webkit-slider-runnable-track):
(.ios .slider &gt; input::-webkit-slider-thumb):
* Source/WebCore/Modules/modern-media-controls/controls/slider-base.js: Added.
(SliderBase.set inputAccessibleLabel):
(SliderBase.prototype.get disabled):
(SliderBase.prototype.set disabled):
(SliderBase.prototype.get value):
(SliderBase.prototype.set value):
(SliderBase.prototype.get secondaryValue):
(SliderBase.prototype.set secondaryValue):
(SliderBase.prototype.get allowsRelativeScrubbing):
(SliderBase.prototype.set allowsRelativeScrubbing):
(SliderBase.prototype.handleEvent):
(SliderBase.prototype.commitProperty):
(SliderBase.prototype._handlePointerdownEvent):
(SliderBase.prototype._interactionEndTarget):
(SliderBase.prototype._valueDidChange):
(SliderBase.prototype._handlePointermoveEvent):
(SliderBase.prototype._handlePointerupEvent):
(SliderBase.prototype._playbackProgress):
* Source/WebCore/Modules/modern-media-controls/controls/slider.css:
(.slider.default &gt; .appearance):
(.slider.default.allows-relative-scrubbing &gt; .appearance.changing):
(.slider.default &gt; .appearance &gt; *):
(.slider.default &gt; .appearance &gt; .fill):
(.slider.default.allows-relative-scrubbing &gt; .appearance &gt; .fill):
(.slider.default &gt; .appearance &gt; .fill.primary):
(.slider.default.allows-relative-scrubbing &gt; .appearance.changing &gt; .fill.primary):
(.slider.default &gt; .appearance &gt; .fill.track):
(.slider.default.allows-relative-scrubbing &gt; .appearance.changing &gt; .fill.track):
(.slider.default &gt; .appearance &gt; .fill.secondary):
(.slider.default.allows-relative-scrubbing &gt; .appearance.changing &gt; .fill.secondary):
(.slider.default &gt; .appearance &gt; .knob):
(.slider.default &gt; .appearance &gt; .knob.circle):
(.slider.default &gt; .appearance &gt; .knob.bar):
(.slider.default &gt; .appearance &gt; .knob.none):
(.slider.default.disabled &gt; .appearance &gt; .primary,):
(.slider): Deleted.
(.slider &gt; .custom-slider,): Deleted.
(.slider &gt; .custom-slider): Deleted.
(.slider.allows-relative-scrubbing &gt; .custom-slider): Deleted.
(.slider.allows-relative-scrubbing &gt; .custom-slider.changing): Deleted.
(.slider &gt; .custom-slider &gt; *): Deleted.
(.slider &gt; .custom-slider &gt; .fill): Deleted.
(.slider.allows-relative-scrubbing &gt; .custom-slider &gt; .fill): Deleted.
(.slider &gt; .custom-slider &gt; .fill.primary): Deleted.
(.slider.allows-relative-scrubbing &gt; .custom-slider.changing &gt; .fill.primary): Deleted.
(.slider &gt; .custom-slider &gt; .fill.track): Deleted.
(.slider.allows-relative-scrubbing &gt; .custom-slider.changing &gt; .fill.track): Deleted.
(.slider &gt; .custom-slider &gt; .fill.secondary): Deleted.
(.slider.allows-relative-scrubbing &gt; .custom-slider.changing &gt; .fill.secondary): Deleted.
(.slider &gt; .custom-slider &gt; .knob): Deleted.
(.slider &gt; .custom-slider &gt; .knob.circle): Deleted.
(.slider &gt; .custom-slider &gt; .knob.bar): Deleted.
(.slider &gt; .custom-slider &gt; .knob.none): Deleted.
(.slider &gt; input): Deleted.
(.slider.allows-relative-scrubbing &gt; input): Deleted.
(.ios .slider &gt; input): Deleted.
(.ios .slider:not(.allows-relative-scrubbing) &gt; input): Deleted.
(.slider &gt; input::-webkit-slider-thumb): Deleted.
(.slider.disabled &gt; input,): Deleted.
(.ios .slider &gt; input::-webkit-slider-runnable-track): Deleted.
(.ios .slider &gt; input::-webkit-slider-thumb): Deleted.
* Source/WebCore/Modules/modern-media-controls/controls/slider.js:
(Slider.prototype.commit):
(Slider.prototype.set inputAccessibleLabel): Deleted.
(Slider.prototype.get disabled): Deleted.
(Slider.prototype.set disabled): Deleted.
(Slider.prototype.get value): Deleted.
(Slider.prototype.set value): Deleted.
(Slider.prototype.get secondaryValue): Deleted.
(Slider.prototype.set secondaryValue): Deleted.
(Slider.prototype.get allowsRelativeScrubbing): Deleted.
(Slider.prototype.set allowsRelativeScrubbing): Deleted.
(Slider.prototype.handleEvent): Deleted.
(Slider.prototype.commitProperty): Deleted.
(Slider.prototype._handlePointerdownEvent): Deleted.
(Slider.prototype._interactionEndTarget): Deleted.
(Slider.prototype._valueDidChange): Deleted.
(Slider.prototype._handlePointermoveEvent): Deleted.
(Slider.prototype._handlePointerupEvent): Deleted.
(Slider.prototype._playbackProgress): Deleted.
* Source/WebCore/Modules/modern-media-controls/js-files:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/251872@main">https://commits.webkit.org/251872@main</a>
</pre>

dcrousso commented on Source/WebCore/Modules/modern-media-controls/controls/abstract-slider.js:
```diff
@@ -0,0 +1,237 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+class AbstractSlider extends LayoutNode
```
NIT: How about `SliderBase`?

Also, if this is something we don't want people to instantiate directly, you could add this
```
// This class should not be instantiated directly. Create a concrete subclass instead.
console.assert(this.constructor !== AbstractSlider && this instanceof AbstractSlider, this);
```

dcrousso commented on Source/WebCore/Modules/modern-media-controls/controls/slider.js:
```diff
@@ -23,94 +23,24 @@
  */
 
-class Slider extends LayoutNode
+class Slider extends AbstractSlider
```
Maybe we call this `BasicSlider` so it's a bit more distinct from `AbstractSlider`/`SliderBase`?

graouts replied:
It's no so much that it's basic so much as it's the default appearance. I'd prefer to leave the class name as-is.

dcrousso commented on Source/WebCore/Modules/modern-media-controls/controls/slider.js:
```diff
@@ -23,94 +23,24 @@
-        super(`<div class="slider ${cssClassName}"></div>`);
-
-        this._layoutDelegate = layoutDelegate;
+        super(layoutDelegate, `default ${cssClassName}`);
```
ditto (:26) with `basic` instead of `default`

dcrousso commented on Source/WebCore/Modules/modern-media-controls/controls/slider.js:
```diff
@@ -130,63 +60,13 @@ class Slider extends LayoutNode
         super.commit();
 
-        let scrubberWidth = (style => {
+        const scrubberWidth = (style => {
```
Why?

graouts replied:
Why not? It's a value that doesn't change :)

dcrousso replied:
as a general FYI, `const` doesn't really work that way in JS :/

all it means is basically "you cannot reassign this identifier (though you can shadow it)".  it doesn't prevent modifications to the value if the value is an object (e.g. `const foo = {}; foo.a = 42;`).

i usually avoid using `const` for anything that isn't an actual constant (e.g. `42`) for this reason (as well as the fact that it's more characters to type).

NIT: my other point here was that these kinds of changes can make it more annoying for a future engineer to go back and look at the `git blame` of lines, as now they have to walk through another commit

graouts replied:
I know `const` isn't similar to how `const` works in some other languages, but here it's appropriate since it _is_ a value that does not change. I personally like to use `const` unless I know a value is _meant_ to change, regardless of whether it is technically possible to change it.

But your point about unnecessary churn stands and I will keep that in mind for future patches.

dcrousso commented on Source/WebCore/Modules/modern-media-controls/controls/slider.js:
```diff
@@ -199,7 +79,7 @@ class Slider extends LayoutNode
         })(this._knobStyle);
 
-        let scrubberBorder = (style => {
+        const scrubberBorder = (style => {
```
ditto (:69)

dcrousso commented on Source/WebCore/Modules/modern-media-controls/controls/slider.js:
```diff
@@ -212,93 +92,14 @@ class Slider extends LayoutNode
         })(this._knobStyle);
 
-        let scrubberCenterX = (scrubberWidth / 2) + Math.round((this.width - scrubberWidth) * this.value);
+        const scrubberCenterX = (scrubberWidth / 2) + Math.round((this.width - scrubberWidth) * this.value);
```
ditto (:69)

dcrousso commented on Source/WebCore/DerivedSources.make:
```diff
@@ -1755,6 +1755,7 @@ POSSIBLE_ADDITIONAL_MODERN_MEDIA_CONTROLS_STYLE_SHEETS = \
             $(ADDITIONS_PATH)/$(STYLE_SHEET)))))
 
 MODERN_MEDIA_CONTROLS_STYLE_SHEETS = \
+    $(WebCore)/Modules/modern-media-controls/controls/abstract-slider.css \
```
I think you need to also add this to `Source/WebCore/CMakeLists.txt` and `Source/WebCore/WebCore.xcodeproj/project.pbxproj`.

dcrousso commented on Source/WebCore/DerivedSources.make:
```diff
@@ -1845,6 +1846,7 @@ MODERN_MEDIA_CONTROLS_SCRIPTS = \
     $(WebCore)/Modules/modern-media-controls/controls/background-tint.js \
     $(WebCore)/Modules/modern-media-controls/controls/time-control.js \
     $(WebCore)/Modules/modern-media-controls/controls/time-label.js \
+    $(WebCore)/Modules/modern-media-controls/controls/abstract-slider.js \
```
ditto (:1758)

dcrousso commented on Source/WebCore/Modules/modern-media-controls/controls/abstract-slider.js:
```diff
@@ -0,0 +1,237 @@
+
+        this._layoutDelegate = layoutDelegate;
+
+        this.appearanceContainer = new LayoutNode(`<div class="slider-appearance"></div>`);
```
So is this basically the same thing as the old `<div class="custom-slider"></div>`?

NIT: Maybe we just do `<div class="appearance"></div>` since we're already inside a `<div class="slider"></div>` (i.e. the `slider-` is kinda redundant)?

graouts replied:
Yes, it was a renaming for something that didn't feel quite obvious. I'll drop the `slider-` prefix!

dcrousso commented on Source/WebCore/Modules/modern-media-controls/controls/abstract-slider.css:
```diff
@@ -0,0 +1,114 @@
+
+.slider.allows-relative-scrubbing > .slider-appearance.changing {
+    top: 0.5px;
+    height: 15px;
```
These make less sense now that the `top` and `height` are not defined in this file.  I'd either move the `.slider.default > .slider-appearance` from the other file back into this one (since it can always be overridden by a more specific selector from a subclass) or move this into the other file so that any `top` and `height` with exact values are always near eachother.

graouts replied:
I've moved these to `slider.css`.

graouts committed:
[media-controls] Allow for Slider subclasses
https://bugs.webkit.org/show_bug.cgi?id=241973

Reviewed by Dean Jackson.

Split the existing code between a SliderBase superclass and a Slider concrete subclass.
SliderBase takes care of defining the model, hosting the <input> and registering the required
DOM events on it, while the subclass takes care of defining its rendering.

* LayoutTests/media/modern-media-controls/resources/media-controls-loader.js:
* LayoutTests/media/modern-media-controls/slider/slider-constructor-expected.txt:
* LayoutTests/media/modern-media-controls/slider/slider-constructor.html:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/modern-media-controls/controls/slider-base.css: Added.
(.slider):
(.slider > .appearance,):
(.slider > .appearance):
(.slider.allows-relative-scrubbing > .appearance):
(.slider.allows-relative-scrubbing > .appearance.changing):
(.slider > input):
(.slider.allows-relative-scrubbing > input):
(.ios .slider > input):
(.ios .slider:not(.allows-relative-scrubbing) > input):
(.slider > input::-webkit-slider-thumb):
(.slider.disabled > input):
(.ios .slider > input::-webkit-slider-runnable-track):
(.ios .slider > input::-webkit-slider-thumb):
* Source/WebCore/Modules/modern-media-controls/controls/slider-base.js: Added.
(SliderBase.set inputAccessibleLabel):
(SliderBase.prototype.get disabled):
(SliderBase.prototype.set disabled):
(SliderBase.prototype.get value):
(SliderBase.prototype.set value):
(SliderBase.prototype.get secondaryValue):
(SliderBase.prototype.set secondaryValue):
(SliderBase.prototype.get allowsRelativeScrubbing):
(SliderBase.prototype.set allowsRelativeScrubbing):
(SliderBase.prototype.handleEvent):
(SliderBase.prototype.commitProperty):
(SliderBase.prototype._handlePointerdownEvent):
(SliderBase.prototype._interactionEndTarget):
(SliderBase.prototype._valueDidChange):
(SliderBase.prototype._handlePointermoveEvent):
(SliderBase.prototype._handlePointerupEvent):
(SliderBase.prototype._playbackProgress):
* Source/WebCore/Modules/modern-media-controls/controls/slider.css:
(.slider.default > .appearance):
(.slider.default.allows-relative-scrubbing > .appearance.changing):
(.slider.default > .appearance > *):
(.slider.default > .appearance > .fill):
(.slider.default.allows-relative-scrubbing > .appearance > .fill):
(.slider.default > .appearance > .fill.primary):
(.slider.default.allows-relative-scrubbing > .appearance.changing > .fill.primary):
(.slider.default > .appearance > .fill.track):
(.slider.default.allows-relative-scrubbing > .appearance.changing > .fill.track):
(.slider.default > .appearance > .fill.secondary):
(.slider.default.allows-relative-scrubbing > .appearance.changing > .fill.secondary):
(.slider.default > .appearance > .knob):
(.slider.default > .appearance > .knob.circle):
(.slider.default > .appearance > .knob.bar):
(.slider.default > .appearance > .knob.none):
(.slider.default.disabled > .appearance > .primary,):
(.slider): Deleted.
(.slider > .custom-slider,): Deleted.
(.slider > .custom-slider): Deleted.
(.slider.allows-relative-scrubbing > .custom-slider): Deleted.
(.slider.allows-relative-scrubbing > .custom-slider.changing): Deleted.
(.slider > .custom-slider > *): Deleted.
(.slider > .custom-slider > .fill): Deleted.
(.slider.allows-relative-scrubbing > .custom-slider > .fill): Deleted.
(.slider > .custom-slider > .fill.primary): Deleted.
(.slider.allows-relative-scrubbing > .custom-slider.changing > .fill.primary): Deleted.
(.slider > .custom-slider > .fill.track): Deleted.
(.slider.allows-relative-scrubbing > .custom-slider.changing > .fill.track): Deleted.
(.slider > .custom-slider > .fill.secondary): Deleted.
(.slider.allows-relative-scrubbing > .custom-slider.changing > .fill.secondary): Deleted.
(.slider > .custom-slider > .knob): Deleted.
(.slider > .custom-slider > .knob.circle): Deleted.
(.slider > .custom-slider > .knob.bar): Deleted.
(.slider > .custom-slider > .knob.none): Deleted.
(.slider > input): Deleted.
(.slider.allows-relative-scrubbing > input): Deleted.
(.ios .slider > input): Deleted.
(.ios .slider:not(.allows-relative-scrubbing) > input): Deleted.
(.slider > input::-webkit-slider-thumb): Deleted.
(.slider.disabled > input,): Deleted.
(.ios .slider > input::-webkit-slider-runnable-track): Deleted.
(.ios .slider > input::-webkit-slider-thumb): Deleted.
* Source/WebCore/Modules/modern-media-controls/controls/slider.js:
(Slider.prototype.commit):
(Slider.prototype.set inputAccessibleLabel): Deleted.
(Slider.prototype.get disabled): Deleted.
(Slider.prototype.set disabled): Deleted.
(Slider.prototype.get value): Deleted.
(Slider.prototype.set value): Deleted.
(Slider.prototype.get secondaryValue): Deleted.
(Slider.prototype.set secondaryValue): Deleted.
(Slider.prototype.get allowsRelativeScrubbing): Deleted.
(Slider.prototype.set allowsRelativeScrubbing): Deleted.
(Slider.prototype.handleEvent): Deleted.
(Slider.prototype.commitProperty): Deleted.
(Slider.prototype._handlePointerdownEvent): Deleted.
(Slider.prototype._interactionEndTarget): Deleted.
(Slider.prototype._valueDidChange): Deleted.
(Slider.prototype._handlePointermoveEvent): Deleted.
(Slider.prototype._handlePointerupEvent): Deleted.
(Slider.prototype._playbackProgress): Deleted.
* Source/WebCore/Modules/modern-media-controls/js-files:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/251872@main
